### DISCUSSION TITLE
Unnecessary duplication of ControlEvents (`<slur>`) in `<app>` and `<choice>`.

### DISCUSSION THREAD

**Discussion by cividini** [music-encoding/music-encoding#1346] [General]
While working on the encoding of Critical Apparatus and Editorial Markup for the Digital Mozart Edition, I encountered a structural problem arising from the "dual" separation of elements in Events and ControlEvents: When ControlEvents refer to Events encapsulated in child elements of an `<app>` (Critical Apparatus) or a `<choice>` (Editorial Markup) element, redundant duplication of the ControlEvent is necessary even if the ControlEvent is not involved in the `<choice>` or `<app>` of the Event.

**Example:** Mozart, KV 165/1, m. 35, Oboe I-II, rhythm correction (Events) without changing the slur (ControlEvent) in Oboe II:

Editor’s correction (`<corr>`):

<img alt="KV_165_01_35_Oboe_corr" src="https://github.com/music-encoding/music-encoding/assets/45197272/8137e186-cc6f-4d3f-928c-6b5d7b69ed05" width="250">

Mozart’s autograph (`<orig>`):

<img alt="KV_165_01_35_Oboe_orig" src="https://github.com/music-encoding/music-encoding/assets/45197272/21a27f82-043f-4d88-901d-db2d2db48549" width="250">

Since the first note in Oboe II (a') requires two different xml:ids for the two child elements `<corr>` and `<orig>` (here in our DME example "note_9714-A" for `<corr>` and "note_9714a-A" with the letter "a" added to the number for `<orig>`), the corresponding `<slur>` must also be duplicated to point to both xml:ids (`@startid="#note_9714-A"` for `<corr>` and `@startid="#note_9714a-A"` for `<orig>`):

<details>
<summary>Click to see the complete MEI file.</summary>
<p>

```xml
<mei meiversion="4.0.1" xmlns="http://www.music-encoding.org/ns/mei">
  <meiHead>
    <fileDesc>
      <titleStmt>
        <title label="NMA digital" type="unit">W. A. Mozart, K 165/1, m. 35, Oboe I-II</title>
      </titleStmt>
      <pubStmt/>
    </fileDesc>
  </meiHead>
  <music>
    <body>
      <mdiv>
        <score>
          <scoreDef key.mode="major" key.pname="f" key.sig="1f" meter.count="4" meter.sym="common" meter.unit="4" optimize="false" xml:id="scoreDef_01-A">
            <staffGrp symbol="bracket" xml:id="staffGrp_02-A">
              <staffDef clef.line="2" clef.shape="G" lines="5" n="1" xml:id="staffDef_P1-A">
                <label xml:id="label_staffP1-A">Oboe I, II</label>
                <labelAbbr xml:id="labelAbbrP1-A">Ob.</labelAbbr>
                <layerDef n="1" xml:id="layerDef_Ob1-A"/>
                <layerDef n="2" xml:id="layerDef_Ob2-A"/>
              </staffDef>
            </staffGrp>
          </scoreDef>
          <section>
            <measure n="35">
              <staff n="1" xml:id="staff_9684-A">
                <layer n="1" xml:id="layer_9687-A">
                  <note dur="4" oct="5" pname="f" tstamp="1" xml:id="note_9690-A"/>
                  <beam xml:id="beam_9693-A">
                    <note dots="1" dur="8" oct="5" pname="c" tstamp="2" xml:id="note_9696-A"/>
                    <note dur="32" oct="5" pname="d" tstamp="2.75" xml:id="note_9699-A"/>
                    <note dur="32" oct="5" pname="c" tstamp="2.875" xml:id="note_9702-A"/>
                  </beam>
                  <note accid.ges="f" dur="4" oct="4" pname="b" tstamp="3" xml:id="note_9705-A"/>
                  <note dur="4" oct="4" pname="a" tstamp="4" xml:id="note_9708-A"/>
                </layer>
                <layer n="2" xml:id="layer_9711-A">
                  <choice xml:id="choice_9720-A">
                    <corr xml:id="corr_9720-A">
                      <note dots="2" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714-A"/>
                      <beam xml:id="beam_9717-A">
                        <note accid.ges="f" dur="32" oct="4" pname="b" tstamp="2.75" xml:id="note_9720-A"/>
                        <note dur="32" oct="4" pname="a" tstamp="2.875" xml:id="note_9723-A"/>
                      </beam>
                    </corr>
                    <orig xml:id="orig_9720-A">
                      <note dots="1" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714a-A"/>
                      <beam xml:id="beam_9717a-A">
                        <note accid.ges="f" dur="16" oct="4" pname="b" tstamp="2.5" xml:id="note_9720a-A"/>
                        <note dur="16" oct="4" pname="a" tstamp="2.75" xml:id="note_9723a-A"/>
                      </beam>
                    </orig>
                  </choice>
                  <note dur="4" oct="4" pname="g" tstamp="3" xml:id="note_9726-A"/>
                  <note dur="4" oct="4" pname="f" tstamp="4" xml:id="note_9729-A"/>
                </layer>
              </staff>
              <slur endid="#note_9708-A" layer="1" staff="1" startid="#note_9696-A" xml:id="slur_9861-A"/>
              <slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714-A" xml:id="slur_9864-A"/>
              <slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714a-A" xml:id="slur_9864a-A"/>
            </measure>
          </section>
        </score>
      </mdiv>
    </body>
  </music>
</mei>
```
</p>
</details>  

The duplication of the slur for `@layer="2"` is obviously redundant, since we have only one slur here for Oboe II. However, the slur would disappear in the rendering of one of the two child elements of `<choice>` if the ControlEvent were not duplicated (here for example in Verovio):

<img alt="KV_165_01_35_Oboe_orig_no_slur" src="https://github.com/music-encoding/music-encoding/assets/45197272/2ec57f91-3f61-4593-bacf-3c812e33ecd3" width="250">

So we are **forced to duplicate the slur**.

The same problem affects `<app>`. 

This is a relevant issue for the encoding of critical editions in MEI, since many variants (`<app>`) or editorial interventions (`<choice>`) only affect events (e.g. notes) without involving ControlEvents (e.g. slurs, trills etc.). So I hope that someone will have the patience to read through this long issue and help find a solution.

Here some possible solutions that are not yet fully satisfactory.

**1. Duplication of the `<choice>` element**

The solution we actually use in the DME is to also encapsulate the ControlEvent in a `<choice>` or `<app>` element:

<details>
<summary>Click to see a partial code of the MEI file.</summary>
<p>

```xml
                <layer n="2" xml:id="layer_9711-A">
                  <choice next="#choice_9864-A" xml:id="choice_9720-A">
                    <corr xml:id="corr_9720-A">
                      <note dots="2" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714-A"/>
                      <beam xml:id="beam_9717-A">
                        <note accid.ges="f" dur="32" oct="4" pname="b" tstamp="2.75" xml:id="note_9720-A"/>
                        <note dur="32" oct="4" pname="a" tstamp="2.875" xml:id="note_9723-A"/>
                      </beam>
                    </corr>
                    <orig xml:id="orig_9720a-A">
                      <note dots="1" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714a-A"/>
                      <beam xml:id="beam_9717a-A">
                        <note accid.ges="f" dur="16" oct="4" pname="b" tstamp="2.5" xml:id="note_9720a-A"/>
                        <note dur="16" oct="4" pname="a" tstamp="2.75" xml:id="note_9723a-A"/>
                      </beam>
                    </orig>
                  </choice>
                  <note dur="4" oct="4" pname="g" tstamp="3" xml:id="note_9726-A"/>
                  <note dur="4" oct="4" pname="f" tstamp="4" xml:id="note_9729-A"/>
                </layer>
                […]
                <choice prev="#choice_9720-A" xml:id="choice_9864-A">
                <corr xml:id="corr_9864-A">
                  <slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714-A" xml:id="slur_9864-A"/>
                </corr>
                <orig xml:id="orig_9864-A">
                  <slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714a-A" xml:id="slur_9864a-A"/>
                </orig>
              </choice>

```
</p>
</details>  

However, this solution is problematic for two reasons:
1. We’re encoding an editorial markup for `<slur>` that doesn’t exist: nothing has been corrected here, there is no difference between "original" and "correction" for the slur. The correction only affects the notes, not the slur.
2. This encoding is more difficult (though not impossible) to process, since the `<choice>` element of the ControlEvent must be ignored when processing the `<choice>` element of the Event. If this doesn’t happen, the ControlEvent will be rendered incorrectly along with the Event (see the unnecessary colour highlighting of the slur in the rendering by MoVi: there is no difference between the two slurs here, it is the same slur, so the slur should not be highlighted):

<img alt="KV_165_01_35_Oboe_orig_slur_marked" src="https://github.com/music-encoding/music-encoding/assets/45197272/c60b8886-4682-4f9f-9064-2db404881b30" width="250">

<img alt="KV_165_01_35_Oboe_corr_slur_marked" src="https://github.com/music-encoding/music-encoding/assets/45197272/ebec3399-627f-4452-9d45-889a4e276d0e" width="250">

(To help the processing one could add e.g. `@type="noshow|hide|secondary"` in the `<choice>` of the ControlEvent `<slur>`, but it would be better to avoid this addition.)

**2. Marking the duplicated ControlEvent as `@sameas` or `@copyof`**

Another possible solution is to mark the duplicated `<slur>` with `@sameas` or `@copyof`:

<details>
<summary>Click to see a partial code of the MEI file.</summary>
<p>

```xml
                <layer n="2" xml:id="layer_9711-A">
                  <choice next="#choice_9864-A" xml:id="choice_9720-A">
                    <corr xml:id="corr_9720-A">
                      <note dots="2" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714-A"/>
                      <beam xml:id="beam_9717-A">
                        <note accid.ges="f" dur="32" oct="4" pname="b" tstamp="2.75" xml:id="note_9720-A"/>
                        <note dur="32" oct="4" pname="a" tstamp="2.875" xml:id="note_9723-A"/>
                      </beam>
                    </corr>
                    <orig xml:id="orig_9720a-A">
                      <note dots="1" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714a-A"/>
                      <beam xml:id="beam_9717a-A">
                        <note accid.ges="f" dur="16" oct="4" pname="b" tstamp="2.5" xml:id="note_9720a-A"/>
                        <note dur="16" oct="4" pname="a" tstamp="2.75" xml:id="note_9723a-A"/>
                      </beam>
                    </orig>
                  </choice>
                  <note dur="4" oct="4" pname="g" tstamp="3" xml:id="note_9726-A"/>
                  <note dur="4" oct="4" pname="f" tstamp="4" xml:id="note_9729-A"/>
                </layer>
[…]
              <slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714-A" xml:id="slur_9864-A"/>
              <slur endid="#note_9729-A" sameas="#slur_9864-A" startid="#note_9714a-A" xml:id="slur_9864a-A"/>

```
</p>
</details>  

`@sameas` seems to me to be better than `@copyof` for its definition ("Points to an element that is the same as the current element but is not a literal copy of the current element."), as the slur with `@sameas` points through the `@startid` to another `@xml:id`, so that the second slur "is not a literary copy of the current element". However, I am not sure if this is an appropriate use of `@sameas` (see the issue https://github.com/music-encoding/music-encoding/issues/1056). And if we consider that the `<slur>` is the same in the same part (`layer="2"`) regardless of the different pointers due to the doubling of the notes in the `<choice>` element, it might be better to use `@copyof` here. Moreover, you need to encode `@startid="#note_9714a-A"` in the "sameas-slur" to make sure that it points to the corresponding note, so that this solution doesn't seem like a real win.

**3. `@tstamp/@tstamp2` instead of `@startid/@endid`**

Using `@tstamp/@tstamp2` instead of `@startid/@endid` seems to be a simple solution to avoid the doubling of the slur: 
```xml
<slur curvedir="below" tstamp="1" layer="2" staff="1" tstamp2="4" xml:id="slur_9864-A"/>
```

This solution seems to me to be too "messy" (regardeless of the inconsistency with the other slurs encoded in the file with `@startid/@endid`) as the slur could hang around somewhere (even if you add `@curvdir`), especially if you change the rendering between two musical readings and if you have different layers in the same staff as in our DME example.

Two other solutions would require changes to the MEI Guidelines:

**4. Duplication of the value in `@startid` and `@endid`**

A small change in MEI would be to allow multiple values in the `@startid` and `@endid` pointers, so that the same slur can point to different notes, in our example two values in the `@startid` for the two notes in `<corr>` and `<orig>`:

```xml
<slur endid="#note_9729-A" layer="2" staff="1" startid="#note_9714-A #note_9714a-A" xml:id="slur_9864-A"/>
```

This solution is however very problematic because it doesn’t fix the order of the two (or more) values in the pointers. This problem becomes clear when we consider the (quite common) case where both notes pointed by the slur have been encoded with different `xml:id`s in `<choice>`:

<img alt="KV_165_01_35_Oboe_corr_begin_end" src="https://github.com/music-encoding/music-encoding/assets/45197272/bddb7cbb-78d4-4b3a-909d-e7b3e9a5933a" width="250">

<details>
<summary>Click to see the complete MEI file.</summary>
<p>

```xml
<mei meiversion="4.0.1" xmlns="http://www.music-encoding.org/ns/mei">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title label="NMA digital" type="unit">W. A. Mozart, K 165/1, m. 35, Oboe I-II</title>
   </titleStmt>
   <pubStmt/>
  </fileDesc>
 </meiHead>
 <music>
  <body>
   <mdiv>
    <score>
     <scoreDef key.mode="major" key.pname="f" key.sig="1f" meter.count="4" meter.sym="common" meter.unit="4" optimize="false" xml:id="scoreDef_01-A">
      <staffGrp symbol="bracket" xml:id="staffGrp_02-A">
       <staffDef clef.line="2" clef.shape="G" lines="5" n="1" xml:id="staffDef_P1-A">
        <label xml:id="label_staffP1-A">Oboe I, II</label>
        <labelAbbr xml:id="labelAbbrP1-A">Ob.</labelAbbr>
        <layerDef n="1" xml:id="layerDef_Ob1-A"/>
        <layerDef n="2" xml:id="layerDef_Ob2-A"/>
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section>
      <measure n="35">
       <staff n="1" xml:id="staff_9684-A">
        <layer n="1" xml:id="layer_9687-A">
         <note dur="4" oct="5" pname="f" tstamp="1" xml:id="note_9690-A"/>
         <beam xml:id="beam_9693-A">
          <note dots="1" dur="8" oct="5" pname="c" tstamp="2" xml:id="note_9696-A"/>
          <note dur="32" oct="5" pname="d" tstamp="2.75" xml:id="note_9699-A"/>
          <note dur="32" oct="5" pname="c" tstamp="2.875" xml:id="note_9702-A"/>
         </beam>
         <note accid.ges="f" dur="4" oct="4" pname="b" tstamp="3" xml:id="note_9705-A"/>
         <note dur="4" oct="4" pname="a" tstamp="4" xml:id="note_9708-A"/>
        </layer>
        <layer n="2" xml:id="layer_9711-A">
         <choice xml:id="choice_9720-A">
          <corr xml:id="corr_9720-A">
           <note dots="2" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714-A"/>
           <beam xml:id="beam_9717-A">
            <note accid.ges="f" dur="32" oct="4" pname="b" tstamp="2.75" xml:id="note_9720-A"/>
            <note dur="32" oct="4" pname="a" tstamp="2.875" xml:id="note_9723-A"/>
           </beam>
          </corr>
          <orig xml:id="orig_9720-A">
           <note dots="1" dur="4" oct="4" pname="a" tstamp="1" xml:id="note_9714a-A"/>
           <beam xml:id="beam_9717a-A">
            <note accid.ges="f" dur="16" oct="4" pname="b" tstamp="2.5" xml:id="note_9720a-A"/>
            <note dur="16" oct="4" pname="a" tstamp="2.75" xml:id="note_9723a-A"/>
           </beam>
          </orig>
         </choice>
         <note dur="4" oct="4" pname="g" tstamp="3" xml:id="note_9726-A"/>
         <choice xml:id="choice_9729-A">
          <corr xml:id="corr_9729-A">
           <note dur="4" oct="4" pname="f" tstamp="4" xml:id="note_9729-A"/>
          </corr>
          <orig xml:id="orig_9729-A">
           <note dur="4" oct="4" pname="d" tstamp="4" xml:id="note_9729a-A"/>
          </orig>
         </choice>
        </layer>
       </staff>
       <slur endid="#note_9708-A" layer="1" staff="1" startid="#note_9696-A" xml:id="slur_9861-A"/>
       <slur endid="#note_9729-A #note_9729a-A" layer="2" staff="1" startid="#note_9714-A #note_9714a-A" xml:id="slur_9864-A"/>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```
</p>
</details>  

The double values in `@startid` and `@endid` of `<slur>` can only be processed if the order of the values is clear: #note_9714-A to #note_9729-A for the slur in `<corr>` or #note_9714a-A to #note_9729a-A for the slur in `<orig>`, BUT NOT #note_9714-A to #note_9729a-A or #note_9714a-A to #note_9729-A.

**5. Duplilcation of the `@startid` and `@endid` attribute**

Another more radical solution which I think would fit good, would be to fix the order of the pointers in `<slur>` through multiple `@startid` and `@endid` attributes, similary to `@tstamp` and `@tstamp2`: `@startid2`, `@startid3`, `@startid4` etc. and `@endid2`, `@endid3`, `@endid4` etc., in our last example:

```xml
<slur endid="#note_9729-A" endid2="#note_9729a-A" layer="2" staff="1" startid="#note_9714-A" startid2="#note_9714a-A" xml:id="slur_9864-A"/>
```

This solution seems to be the more consistent one, but I see a very radical modification of MEI here, as we would allow multiple `@startid` and `@endid` attributes. In the case of an `<app>` with many variants (`<rdg1><rdg2><rdg3>` etc.) we would need multiple `@startid` and/or `@endid` for multiple variants: `@startid`, `@startid2`, `@ startid3` and/or `@endid`, `@endid2`, `@endid3` etc. in a single `<slur>`). 

I look forward to any comments, criticisms and suggestions!

**Comment by ahankinson** [comment]
Just a quick note: Duplicate attributes on the same element is not valid XML.

**Comment by rettinghaus** [comment]
The slur "looks" the same, but the start notes are not the same notes, so I'd say they are actually different slurs (so using `@sameas` or `@copyof` would be just wrong). Even the slightest difference justifies the "double-encoding" of elements, say in the autograph the composer used red or blue ink for a slur and in print it is black. 
What is the actual problem with this?

**Comment by ahankinson** [comment]
To me, options 2 and 3 are the best.

Option 1 is problematic for the reasons you outline.

Option 4 is problematic because multiple attribute values are not understood as a choice, but as a requirement. That is, any processor must try to resolve both. It will also cause problems with renderers since the slur will point to an ID that will not exist in the DOM. 

Option 5 is not valid XML as I said.

**Comment by lpugin** [comment]
I am suggesting to make this a discussion. Anybody would prefer not to?
